### DISCUSSION TITLE
Are there any Livewire hooks or events that I can access before a network request is sent?

### DISCUSSION THREAD

**Discussion by citricguy** [livewire/livewire#5128] [1. Help] [answered]
Are there any available hooks/events on the client side of Livewire that can be tapped into just before a network request is sent? 

What I'm hoping to be able to do is check for the availability of a CSRF-Token before any requests are sent, so that if it is not available I can grab a new CSRF token to prevent the inevitable 419 response. My users take issue with the built in warn-and-refresh approach.

I can get everything working, I just don't know how/where to trigger things 'before' the Livewire requests are sent.
```
async function updateXSRF() {
    if (document.cookie.indexOf('XSRF-TOKEN=') == -1) {
        console.log('New CSRF Token needed!');
        await fetch('/csrf-token').then((response) => {
            return response.json();
        }).then((data) => {
            console.log('Got new token: ' + data.csrfToken);
            document.getElementsByTagName('meta')["CSRF-Token"].content = data.csrfToken;
        })
    } else {
        console.log('Cookie Still Available');
    }
}
```

For interest sake, here is what the /csrf-token end point returns from the above example
`return ['csrfToken' => csrf_token()];`

**Comment by joshhanley** [comment]
@citricguy the only one I know of that is close is `message.sent` https://laravel-livewire.com/docs/2.x/lifecycle-hooks#js-hooks

But we did add an option for customising the 419, so instead of it showing the `alert()` you can customise it to request a new csrf token.

To do this, I think you need to register a custom callback by passing it into 

```js
Livewire.onPageExpired((response, message) => {
    // Do stuff
})
```

https://github.com/livewire/livewire/blob/020ad095cf1239138b097d22b584e2701ec3edfb/js/index.js#L115-L117

Which then gets used here 

https://github.com/livewire/livewire/blob/020ad095cf1239138b097d22b584e2701ec3edfb/js/connection/index.js#L20-L28

Hope this helps!

**Reply by citricguy** [reply]
Thank you, this gets me closer. Now I'm looking into the possibility of being able to 'resend' the last request automatically after refreshing an expired token, similar to how [Axios interceptors](https://dev.to/grantholle/better-csrf-refreshing-in-laravel-and-axios-177c) work for this issue with Inertia JS.

**Reply by joshhanley** [reply]
No worries!

**Comment by Martleby** [comment, marked as answer]
Here's a thought...

I'm guessing you have a button hooked up with a wire:click="myComponentMethod" or similar? 

If you're using Alpine you could replace this with an @click and call an Alpine method which would then do your csrf check before issuing $wire.call('myComponentMethod', ...args).

In vanilla JS, you could create a regular onclick handler which performs your check and then calls the component method.  To do that you could use @this.call('myComponentMethod', ...args) - @this is special blade directive that will get the component instance for you.

**Reply by citricguy** [reply]
Alpine is a great idea! I believe I'll start with this until I can figure out how to permanently handle this 419 error issue.
